### PR TITLE
Broker contract

### PR DESCRIPTION
**Pull request by ninabarbakadze**
### Description

Implements the Broker contract following IBroker and IBrokerAdmin interfaces.

### Tested

Unit Tests

### CODE REVIEW

**Review by ninabarbakadze — COMMENTED** [review]
It's still a WIP so NatSpec will have to be cleaned up a bit more

File: packages/protocol/contracts/stability/Broker.sol [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+/**
+ * @title Broker
+ * @notice The broker executes swaps and keeps track of spending limits per pair
+ */
+contract Broker is IBroker, IBrokerAdmin, Initializable, Ownable {
+  /* ==================== State Variables ==================== */
+  address[] exchangeManagers;
+
+  // Address of the broker contract.
+  IExchangeManager public exchangeManager;
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: This should be removed as the contract now manages multiple exchangeManagers.

File: packages/protocol/contracts/stability/Broker.sol [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import { IReserve } from "./interfaces/IReserve.sol";
+import { Initializable } from "../common/Initializable.sol";
+
+/**
+ * @title Broker
+ * @notice The broker executes swaps and keeps track of spending limits per pair
+ */
+contract Broker is IBroker, IBrokerAdmin, Initializable, Ownable {
+  /* ==================== State Variables ==================== */
+  address[] exchangeManagers;
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: We should also keep a mapping of `address => bool` to check for inclusion.

File: packages/protocol/contracts/stability/Broker.sol [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Allows the contract to be upgradable via the proxy.
+   * @param _exchangeManager The address of the PairManager contract.
+   * @param _reserve The address of the Rezerve contract.
+   */
+  function initilize(address _exchangeManager, address _reserve) external initializer {
+    _transferOwnership(msg.sender);
+    addExchangeManager(_exchangeManager);
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: Should be initialized with an array of exchangeManagers.

File: packages/protocol/contracts/stability/Broker.sol [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountOut The amount of tokenOut to be bought
+   * @return amountIn The amount of tokenIn to be sold
+   */
+  function getAmountIn(bytes32 exchangeId, address tokenIn, address tokenOut, uint256 amountOut)
+    external
+    returns (uint256 amountIn)
+  {
+    return amountOut = exchangeManager.getAmountIn(exchangeId, tokenIn, tokenOut, amountOut);
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: This function should also take in an `address exchangeManager` as the first parameter, check if it's a registered exchangeManager and then cast that to an `IExchangeManager(exchangeManager)` and call getAmountIn on it.

File: packages/protocol/contracts/stability/Broker.sol [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought 
+   * @param amountIn The amount of tokenIn to be sold
+   * @return amountOut The amount of tokenOut to be bought
+   */
+  function getAmountOut(bytes32 exchangeId, address tokenIn, address tokenOut, uint256 amountIn)
+    external
+    returns (uint256 amountOut)
+  {
+    return amountOut = exchangeManager.getAmountOut(exchangeId, tokenIn, tokenOut, amountIn);
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: Same as above

File: packages/protocol/contracts/stability/Broker.sol [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn,
+    uint256 amountOutMin
+  ) external returns (uint256 amountOut) {
+    require(existsInExchangeManagers(_exchangeManager), "ExchangeManager does not exist");
+    exchangeManager = IExchangeManager(_exchangeManager);
+    return
+      amountOut = exchangeManager.swapIn(exchangeId, tokenIn, tokenOut, amountIn, amountOutMin);
+    emit Swap(_exchangeManager, exchangeId, msg.sender, tokenIn, tokenOut, amountIn, amountOut);
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: Ok you did it here, the same logic should happen in getAmountIn/Out as well.

File: packages/protocol/contracts/stability/Broker.sol [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+   * @return bool Returns true or false
+   */
+  function existsInExchangeManagers(address _exchangeManager) public view returns (bool) {
+    for (uint256 i = 0; i < exchangeManagers.length; i++) {
+      if (exchangeManagers[i] == _exchangeManager) {
+        return true;
+      }
+    }
+    return false;
+  }
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: As per above, we should keep a mapping for this check.

**Review by bowd — CHANGES_REQUESTED** [review]

File: packages/protocol/contracts/stability/Broker.sol [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+   * @param test Set to true to skip implementation initialization.
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Allows the contract to be upgradable via the proxy.
+   * @param _exchangeManagers The addresses of the ExchangeManager contracts.
+   * @param _reserve The address of the Reserve contract.
+   */
+  function initilize(address[] calldata _exchangeManagers, address _reserve) external initializer {
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: typo `initilize` -> `initialize`

File: packages/protocol/contracts/stability/Broker.sol [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+   * @return bool Returns true or false
+   */
+  function checkIsExchangeManager(address exchangeManager) public view returns (bool) {
+    for (uint256 i = 0; i < exchangeManagers.length; i++) {
+      if (exchangeManagers[i] == exchangeManager && isExchangeManager[exchangeManagers[i]]) {
+        return true;
+      }
+    }
+    return false;
+  }
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: We can get rid of this function completely, the mapping being public exposes a view function `isExchangeManager(address)` which accomplishes the same thing.

File: packages/protocol/contracts/stability/Broker.sol [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+  /* ==================== Mutative Functions ==================== */
+
+  /**
+   * @notice Add exchange manager
+   * @param exchangeManager The address of the exchange manager to add
+   * @return index The index where it was inserted
+   */
+  function addExchangeManager(address exchangeManager) public onlyOwner returns (uint256 index) {
+    require(!checkIsExchangeManager(exchangeManager), "ExchangeManager already exists in the list");
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: This and all following `checkIsExchangeManager(exchangeManager)` should be `isExchangeManager[exchangeManager]`.

**Review by bowd — CHANGES_REQUESTED** [review]

File: packages/protocol/contracts/stability/Broker.sol [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+  ) external returns (uint256 amountOut) {
+    require(checkIsExchangeManager(exchangeManager), "ExchangeManager does not exist");
+    return
+      amountOut = IExchangeManager(exchangeManager).swapIn(
+        exchangeId,
+        tokenIn,
+        tokenOut,
+        amountIn,
+        amountOutMin
+      );
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: There's one more thing here. I know this isn't currently reflected in the interface but we decided to do slippage control on the Broker rather than the exchange manager which means that I'd need you to do two things on your branch:

1. Edit `IExchangeManager.sol` and remove the `amountOutMin` and `amountInMax` from the `swapIn` and `swapOut` functions.
2. Here add an additional require for `amountOut >= amountOutMin` and revert with "amountOutMin not met" and similarly in `swapOut` have `amountIn <= amountInMax` and revert with "amountInMax exceeded"

Makes sense?

Comment (ninabarbakadze) [reply packages/protocol/contracts/stability/Broker.sol]: yes that was done in the order version

**Review by bowd — CHANGES_REQUESTED** [review]
Looking good, now how about some tests? :D

File: packages/protocol/contracts/stability/Broker.sol [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+  function swapIn(
+    address exchangeManager,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn,
+    uint256 amountOutMin
+  ) external returns (uint256 amountOut) {
+    require(isExchangeManager[exchangeManager], "ExchangeManager does not exist");
+    amountOut = IExchangeManager(exchangeManager).swapIn(exchangeId, tokenIn, tokenOut, amountIn);
```
Comment (bowd) [inline packages/protocol/contracts/stability/Broker.sol]: So my brain was a bit absent on Friday, and I realized that I forgot the most important thing 😂 
The Broker's main responsibility is to actually execute the token transfers/mint/burn involved in a swap.
The `IExchangeManager`'s only responsible for pricing things. Check [this section](https://docs.google.com/document/d/1fII7pHpBqnlPtDXj0JdvAIAr6V5JZN5ZuJREYYCaww8/edit#heading=h.wquz253tdlws) of the design doc.

So the broker should have two internal functions:
1. `transferOut(address to, address token,  uint256 amount)` 
2. `transferIn(address token, uint256 amount)`

Each will have to split logic based on what type of token they're dealing with, like so:

**transferOut**

- if `Reserve.isStableAsset(token)` then `StableToken(token).mint(to, amount)`.
- if `Reserve.isCollateralAsset(token)` then `Reserve.transferExchangeCollateralAsset(token, to, amount)`

**transferIn**

- if `Reserve.isStableAsset(token) then `StableToen(token).transferFrom(to, address(this), amount)` && `StableToken(token).burn(amount)`
- if `Reserve.isCollateralAsset(token)` then `IERC20(token).transferFrom(to, address(reserve), amount)`

---

Then basically you can use the functions in `swapIn` and `swapOut`, to execute the trade `transferIn(tokenIn, amountIn)` and `transferOut(tokenOut, amountOut)`.

Makes sense?

File: packages/protocol/contracts/stability/Broker.sol [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+   * @notice This method is responsible for burning assets
+   * @param token The asset getting burned
+   * @param amount The amount of asset getting burned
+   */
+  function transferIn(address token, uint256 amount) private {
+    if (reserve.isStableAsset(token)) {
+      IERC20(token).transferFrom(msg.sender, address(this), amount);
+      IStableToken(token).burn(amount);
+    }
+    if (reserve.isCollateralAsset(token)) {
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: Use else-if instead here

File: packages/protocol/test-sol/stability/Broker.t.sol [hunk 14]
```diff
@@ -114,8 +74,8 @@ contract BrokerTest_initilizerAndSetters is BrokerTest {
     assertEq(broker.owner(), deployer);
   }
 
-  function test_initilize_shouldSetPairManager() public {
-    assertEq(address(broker.pairManager()), address(pairManager));
+  function test_initilize_shouldSetExchangeProviderAddresseses() public {
+    assertEq(broker.getExchangeProviders(), broker.getExchangeProviders());
```
Comment (bayological) [inline packages/protocol/test-sol/stability/Broker.t.sol]: This looks like a mistake. We should probably be comparing the list that was passed into the initilizer with the list returned from getExchangeProviders

File: packages/protocol/contracts/stability/Broker.sol [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+
+  /* ==================== Mutative Functions ==================== */
+
+  /**
+   * @notice Add exchange manager
+   * @param exchangeProvider The address of the exchange manager to add
+   * @return index The index where it was inserted
+   */
+  function addExchangeProvider(address exchangeProvider) public onlyOwner returns (uint256 index) {
+    require(!isExchangeProvider[exchangeProvider], "ExchangeProvider already exists in the list");
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol:60]: There's currently no unit test to verify this condition. We need to have one that checks this works as expected.

File: packages/protocol/contracts/stability/Broker.sol [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+    require(exchangeProvider != address(0), "ExchangeProvider address can't be 0");
+    exchangeProviders.push(exchangeProvider);
+    isExchangeProvider[exchangeProvider] = true;
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol:63]: There's nowhere in the tests where we are checking this. Each write to storage should be checked. Please update the happy path test to verify this has been set. I would verify it was false before calling the function then verify it is true after.

File: packages/protocol/test-sol/stability/Broker.t.sol [hunk 17]
```diff
@@ -124,26 +84,52 @@ contract BrokerTest_initilizerAndSetters is BrokerTest {
...
-    vm.expectRevert("PairManager address must be set");
-    broker.setPairManager(address(0));
+    vm.expectRevert("ExchangeProvider address can't be 0");
+    broker.addExchangeProvider(address(0));
   }
 
-  function test_setPairManager_whenSenderIsOwner_shouldUpdateAndEmit() public {
+  function test_addExchangeProvider_whenSenderIsOwner_shouldUpdateAndEmit() public {
     changePrank(deployer);
     address newPairManager = actor("newPairManager");
```
Comment (bayological) [inline packages/protocol/test-sol/stability/Broker.t.sol]: This variable name needs to be updated along with the usages.

File: packages/protocol/contracts/stability/Broker.sol [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+   * @param exchangeProvider The address of the exchange provider to add.
+   * @return index The index of the newly added specified exchange provider.
+   */
+  function addExchangeProvider(address exchangeProvider) public onlyOwner returns (uint256 index) {
+    require(!isExchangeProvider[exchangeProvider], "ExchangeProvider already exists in the list");
+    require(exchangeProvider != address(0), "ExchangeProvider address can't be 0");
+    exchangeProviders.push(exchangeProvider);
+    isExchangeProvider[exchangeProvider] = true;
+    emit ExchangeProviderAdded(exchangeProvider);
+    return index = exchangeProviders.length - 1;
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: There's no need to explicitly call return when you've named the return variable in the function signature.

File: packages/protocol/contracts/stability/Broker.sol [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+
+  /**
+   * @notice Remove an exchange manager at an index
+   * @param exchangeProvider The address of the exchange manager to remove
+   * @param index The index in the exchange managers array
+   */
+  function removeExchangeProvider(address exchangeProvider, uint256 index) public onlyOwner {
+    require(
+      index < exchangeProviders.length && exchangeProviders[index] == exchangeProvider,
+      "index into exchangeProviders list not mapped to token"
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: This error message could be made clearer. What is "token"? And also consider splitting the checks into separate require statements so you can be specific about the failure reason for each condition.

**Review by bayological — CHANGES_REQUESTED** [review]

File: packages/protocol/contracts/stability/Broker.sol [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+   * @param _reserve The Mento reserve address
+   */
+  function setReserve(address _reserve) public onlyOwner {
+    require(_reserve != address(0), "Reserve address must be set");
+    emit ReserveSet(_reserve, address(reserve));
+    reserve = IReserve(_reserve);
+  }
+
+  /**
+   * @notice Calculate amountIn of tokenIn for a given amountIn of tokenIn
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: This should be rewritten, so it makes more sense in plain English. After reading this I don't understand what this function  is supposed to be doing.

File: packages/protocol/contracts/stability/Broker.sol [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+  function getAmountIn(
+    address exchangeProvider,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountOut
+  ) external returns (uint256 amountIn) {
+    require(isExchangeProvider[exchangeProvider], "ExchangeProvider does not exist");
+    return
+      amountIn = IExchangeProvider(exchangeProvider).getAmountIn(
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: If you're using named return variables you only need to do the assignment, you don't need to call return.

File: packages/protocol/contracts/stability/Broker.sol [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+  function getAmountOut(
+    address exchangeProvider,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn
+  ) external returns (uint256 amountOut) {
+    require(isExchangeProvider[exchangeProvider], "ExchangeProvider does not exist");
+    return
+      amountOut = IExchangeProvider(exchangeProvider).getAmountOut(
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: Same as above, If you're using named return variables you only need to do the assignment, you don't need to call return.

File: packages/protocol/contracts/stability/Broker.sol [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+import { Initializable } from "../common/Initializable.sol";
+
+/**
+ * @title Broker
+ * @notice The broker executes swaps and keeps track of spending limits per pair
+ */
+contract Broker is IBroker, IBrokerAdmin, Initializable, Ownable {
+  /* ==================== State Variables ==================== */
+
+  address[] exchangeProviders;
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: We should explicitly set the visibility for variables.

File: packages/protocol/contracts/stability/Broker.sol [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+    require(amountIn <= amountInMax, "amountInMax exceeded");
+    transferIn(msg.sender, tokenIn, amountIn);
+    transferOut(msg.sender, tokenOut, amountOut);
+    emit Swap(exchangeProvider, exchangeId, msg.sender, tokenIn, tokenOut, amountIn, amountOut);
+  }
+
+  /* ==================== Private Functions ==================== */
+
+  /**
+   * @notice This method is responsible for minting tokens
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: This needs a more detailed explanation. The function doesn't just mint tokens; it also transfers reserve collateral.

File: packages/protocol/contracts/stability/Broker.sol [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+      IStableToken(token).mint(to, amount);
+    } else if (reserve.isCollateralAsset(token)) {
+      reserve.transferCollateralAsset(token, to, amount);
+    } else {
+      revert("Unexpected token type");
+    }
+  }
+
+  /**
+   * @notice This method is responsible for burning assets
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: Same as above. Let's get a more detailed explanation added of what this function is supposed to be doing.

**Review by bayological — CHANGES_REQUESTED** [review]

File: packages/protocol/contracts/stability/Broker.sol [hunk 26]
```diff
@@ -0,0 +1,248 @@
...
+   * @param token The asset getting minted
+   * @param amount The amount of asset getting minted
+   */
+  function transferOut(address payable to, address token, uint256 amount) internal {
+    if (reserve.isStableAsset(token)) {
+      IStableToken(token).mint(to, amount);
+    } else if (reserve.isCollateralAsset(token)) {
+      reserve.transferCollateralAsset(token, to, amount);
+    } else {
+      revert("Unexpected token type");
```
Comment (bayological) [inline packages/protocol/contracts/stability/Broker.sol]: Let's be more specific with the revert messages. Something like "Token is not stable or collateral asset" | "Token must be stable or collateral assert". When writing error messages, we should try to keep them short, but they should also be as clear as possible.

File: packages/protocol/test-sol/stability/Broker.t.sol [hunk 27]
```diff
@@ -162,226 +148,339 @@ contract BrokerTest_initilizerAndSetters is BrokerTest {
...
+    assertEq(IERC20(collateralAsset).balanceOf(address(reserve)), 0);
   }
 
-  function test_swap_minAmountNotMet_shouldRevert() public {
-    bytes32 pairId = mockGetPair();
-    uint256 amountIn = 1e18;
-    uint256 mockAmountOut = 2e18;
-    uint256 nextStableBucket = initialStableBucket + amountIn;
-    uint256 nextCollateralBucket = initialCollateralBucket - mockAmountOut;
+  function test_swapOut_tokenInCollateralAsset_shouldSwap() public {
```
Comment (bayological) [inline packages/protocol/test-sol/stability/Broker.t.sol]: Test naming convention.

File: packages/protocol/test-sol/stability/Broker.t.sol [hunk 28]
```diff
@@ -162,226 +148,339 @@ contract BrokerTest_initilizerAndSetters is BrokerTest {
...
 
-    vm.expectCall(
-      address(pairManager),
+    assertEq(1e16, amountOut);
+    assertEq(IERC20(collateralAsset).balanceOf(trader), 0);
+    assertEq(IERC20(collateralAsset).balanceOf(address(reserve)), 2e16);
+    assertEq(stableAsset.balanceOf(trader), 2e16);
+  }
+
+  function test_swapOut_tokenInStableAsset() public {
```
Comment (bayological) [inline packages/protocol/test-sol/stability/Broker.t.sol]: Test naming convention.

File: packages/protocol/test-sol/stability/Broker.t.sol [hunk 29]
```diff
@@ -162,226 +148,339 @@ contract BrokerTest_initilizerAndSetters is BrokerTest {
...
+      address(exchangeProvider),
+      exchangeId,
+      address(stableAsset),
+      address(collateralAsset),
+      amountOut,
+      1e15
     );
+  }
+
+  function test_swapIs_tokenInStableAsset_shouldSwap() public {
```
Comment (bayological) [inline packages/protocol/test-sol/stability/Broker.t.sol]: Typo in the name & Test naming convention.

File: packages/protocol/test-sol/stability/Broker.t.sol [hunk 30]
```diff
@@ -162,226 +148,339 @@ contract BrokerTest_initilizerAndSetters is BrokerTest {
...
+    broker.getAmountIn(
+      randomExchangeProvider,
+      exchangeId,
+      address(stableAsset),
+      address(collateralAsset),
+      1e24
+    );
+  }
+
+  function test_getAmountIn_receivedCall() public {
```
Comment (bayological) [inline packages/protocol/test-sol/stability/Broker.t.sol]: Naming convention.

File: packages/protocol/contracts/stability/Broker.sol (hunk 19 above)
Comment (ninabarbakadze) [reply packages/protocol/contracts/stability/Broker.sol]: I'd keep the check as is since both conditions are checking whether the index is out of range

Comment (ninabarbakadze) [reply packages/protocol/contracts/stability/Broker.sol]: "token" is by mistake should say exchangeProvider

File: packages/protocol/test-sol/stability/Broker.t.sol (hunk 27 above)
Comment (ninabarbakadze) [reply packages/protocol/test-sol/stability/Broker.t.sol]: how do you mean?

File: packages/protocol/contracts/stability/Broker.sol (hunk 19 above)
Comment (bayological) [reply packages/protocol/contracts/stability/Broker.sol]: Yeah so my point wasn't to remove the check, it was to separate them so it's clear why it fails. But looking at it now, you actually only need the second check `exchangeProviders[index] == exchangeProvider.`

File: packages/protocol/test-sol/stability/Broker.t.sol (hunk 27 above)
Comment (bayological) [reply packages/protocol/test-sol/stability/Broker.t.sol]: *test_swapOut_whenTokenInCollateralAsset_shouldSwap()
